### ISSUE TITLE
How to add metrics if I don't know they already exist or not?

### ISSUE THREAD

**Issue by jdx-john** [jupp0r/prometheus-cpp#451]
I have a class who registers a metric:

MyConstructor(string name)
{
        BuildCounter()
        .Name("time_running_seconds_total")
        .Help("How many seconds is this server running?")
        .Labels({ {"main_label", name} })
        .Register();
}

Normally there is only one instance, but in one application several are created and run in parallel. I realised this is the source of my problem... `throw std::invalid_argument("Family name already exists");`

How can/should I cope with this situation? I don't want the metric to be created at a higher level in the application, ideally - I want each class to create the metrics it reports. I had assumed they would get merged and there is Merge functionality but it doesn't appear to work. I just want to ensure the metrics with the correct dimensions exist, without checking... have I misunderstood? Is there a workaround?

```
template <typename T>
Family<T>& Registry::Add(const std::string& name, const std::string& help,
                         const std::map<std::string, std::string>& labels) {
  std::lock_guard<std::mutex> lock{mutex_};

  if (NameExistsInOtherType<T>(name)) {
    throw std::invalid_argument(
        "Family name already exists with different type");
  }

  auto& families = GetFamilies<T>();

  if (insert_behavior_ == InsertBehavior::Merge) {
    auto same_name_and_labels =
        [&name, &labels](const std::unique_ptr<Family<T>>& family) {
          return std::tie(name, labels) ==
                 std::tie(family->GetName(), family->GetConstantLabels());
        };

    auto it =
        std::find_if(families.begin(), families.end(), same_name_and_labels);
    if (it != families.end()) {
      return **it;
    }
  }

  if (insert_behavior_ != InsertBehavior::NonStandardAppend) {
    auto same_name = [&name](const std::unique_ptr<Family<T>>& family) {
      return name == family->GetName();
    };

    auto it = std::find_if(families.begin(), families.end(), same_name);
    if (it != families.end()) {
      throw std::invalid_argument("Family name already exists");
    }
  }

  auto family = detail::make_unique<Family<T>>(name, help, labels);
  auto& ref = *family;
  families.push_back(std::move(family));
  return ref;
}
```

**Comment by gjasny** [comment]
Hello,

If you want to have the newly added Familiy merged into the a pre-existing one both need to have the same constant labels.
In case you need different labels, please assign them while building the actual metric types with the `Builder`.

I know the merging strategy could be much smarter but I simply lack the time to re-fine the implementation right now.

Thanks,
Gregor

**Comment by jdx-john** [comment]
Thanks @gjasny this makes sense and works fine. Closing :)
